### ISSUE TITLE
bug(training): SurgeXTDataset opens HDF5 handle in __init__ — fork-unsafe for multi-GPU DDP + num_workers>0

### ISSUE THREAD

**Issue by ktinubu** [tinaudio/synth-setter#247]
## Summary

`SurgeXTDataset.__init__` opens an `h5py.File` handle eagerly at construction time (`self.dataset_file = h5py.File(dataset_file, "r")`). When used with PyTorch Lightning DDP (multi-GPU) and `num_workers > 0`, forked dataloader worker processes inherit this handle, causing **silent data corruption, segfaults, or deadlocks** due to HDF5's internal state being shared across forks.

Additionally, `ShiftedBatchSampler` is not distributed-aware — all DDP ranks generate identical permutations, meaning every GPU trains on the **same data** (wasting all but one GPU).

## Affected file

`src/data/surge_datamodule.py`

## Current behavior (with `devices=4, num_workers=8`)

1. **HDF5 fork corruption**: `setup()` creates `SurgeXTDataset` instances that open file handles. DDP spawns 4 rank processes, then each rank's DataLoader forks 8 workers → 32 processes sharing 4 HDF5 file descriptors. Results in:
   - **Segfault** — HDF5 internal state mangled across forks (most common)
   - **Silent wrong data** — reads return bytes from wrong offsets; loss looks normal but model trains on garbage (worst case, undetectable without explicit verification)
   - **Deadlock** — HDF5 internal locks deadlock across forked processes; training freezes at 0% GPU utilization
2. **Duplicate data across ranks**: `ShiftedBatchSampler` has no `rank`/`num_replicas` awareness. All 4 GPUs iterate the same permutation of batches. With `batch_size=None`, Lightning skips `DistributedSampler` auto-wrapping, so there's no safety net. Training appears to converge normally but 3 of 4 GPUs are wasted.
3. **`teardown()` crashes after fix**: If the handle is made lazy, `teardown()` calls `self.train_dataset.dataset_file.close()` on an attribute that no longer exists.

## Root cause

```python
# surge_datamodule.py line ~47
self.dataset_file = h5py.File(dataset_file, "r")  # opened in __init__, survives fork()
```

HDF5 file handles are not fork-safe. When `os.fork()` is called (by DDP or DataLoader workers), the child inherits the parent's file descriptor and HDF5 internal state. Two processes then share the same low-level file descriptor with independent read cursors, corrupting each other's reads.

## Expected behavior

- Each forked worker process opens its own independent HDF5 handle
- Each DDP rank trains on a non-overlapping partition of the dataset
- `teardown()` safely closes whatever handles are open

### 1. Lazy file handle with PID check

```python
def __init__(self, dataset_file, batch_size, ...):
    self.dataset_path = str(dataset_file)
    self._file = None
    self._pid = None

    # Read length once, then close
    with h5py.File(dataset_file, "r") as f:
        self._audio_len = f["audio"].shape[0]

    if use_saved_mean_and_variance:
        self._load_dataset_statistics(dataset_file)

def _open(self):
    pid = os.getpid()
    if self._file is None or self._pid != pid:
        self._file = h5py.File(self.dataset_path, "r")
        self._pid = pid

def __getitem__(self, idx):
    if self.fake:
        return self._get_fake_item()
    self._open()
    # ... rest uses self._file instead of self.dataset_file

def __len__(self):
    if self.fake:
        return 10000
    return self._audio_len // self.batch_size
```

### 2. Distributed-aware `ShiftedBatchSampler`

```python
import torch.distributed as dist

class ShiftedBatchSampler(torch.utils.data.BatchSampler):
    def __init__(self, batch_size, num_batches, num_replicas=None, rank=None):
        self.batch_size = batch_size
        self.num_batches = num_batches
        if num_replicas is None:
            num_replicas = dist.get_world_size() if dist.is_initialized() else 1
        if rank is None:
            rank = dist.get_rank() if dist.is_initialized() else 0
        self.num_replicas = num_replicas
        self.rank = rank

    def __len__(self):
        return (self.num_batches - 1) // self.num_replicas

    def __iter__(self):
        offset = random.randint(0, self.batch_size - 1)
        perm = np.random.permutation(self.num_batches - 1)
        perm = perm[self.rank::self.num_replicas]
        for i in perm:
            yield (i * self.batch_size + offset, (i + 1) * self.batch_size + offset)
```

### 3. Fix `teardown()`

```python
def teardown(self, stage=None):
    for ds in [self.train_dataset, self.val_dataset, self.test_dataset]:
        if hasattr(ds, '_file') and ds._file is not None:
            ds._file.close()
```

## Testing strategy

1. **Fork-safety unit test**: Open dataset in parent, fork a child via `multiprocessing.Process`, read same index in both → assert values match
2. **Multi-worker consistency test**: Read samples directly vs. through `DataLoader(num_workers=2)` → assert identical
3. **Distributed sampler non-overlap test**: Instantiate `ShiftedBatchSampler` for 4 ranks → assert no sample index overlap and >90% coverage
4. **Lazy open test**: Assert `_file is None` after `__init__`, assert `_file is not None` after first `__getitem__`

## Alternative: migrate to WebDataset

A more comprehensive solution is to replace the HDF5 VDS + custom sampler approach with [WebDataset](https://github.com/webdataset/webdataset), which handles fork-safety, distributed sharding, and streaming I/O out of the box. The existing 200-shard structure maps naturally to WebDataset tar shards. This would eliminate all custom sampler code and the HDF5 concurrency issues entirely. The tradeoff is losing the shifted-batch sampling strategy and requiring a one-time shard conversion.

## Environment

- PyTorch Lightning with DDP strategy
- `num_workers=8` (passed via Hydra)
- `batch_size=128` (passed via Hydra)
- 200 HDF5 shards behind a virtual dataset (VDS)
- Target: 4-GPU training
